File: app/src/main/java/org/docheinstein/minimote/hotkeys/HotkeysFragment.java
package org.docheinstein.minimote.hotkeys;

import android.content.ClipData;
import android.content.ClipDescription;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.StateListDrawable;
import android.os.Bundle;
import android.util.Log;
import android.view.DragEvent;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.Observer;
import androidx.navigation.fragment.NavHostFragment;

import org.docheinstein.minimote.R;
import org.docheinstein.minimote.base.MinimoteFragment;
import org.docheinstein.minimote.database.DB;
import org.docheinstein.minimote.database.hotkey.HotkeyEntity;
import org.docheinstein.minimote.settings.SettingsManager;

import java.util.List;

public class HotkeysFragment extends MinimoteFragment {
    private static final String TAG = "HotkeysFragment";

    private Bundle mSavedState;
    private FrameLayout uiHotkeysContainer;

    private MenuItem mSaveButton;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        Log.d(TAG, "onCreateView");

        if(savedInstanceState != null && mSavedState == null) {
            mSavedState = savedInstanceState.getBundle(TAG);
        }

        View view = inflater.inflate(R.layout.hotkeys, container, false);

        uiHotkeysContainer = view.findViewById(R.id.uiHotkeysContainer);
        uiHotkeysContainer.setBackgroundColor(SettingsManager.hotkeysOverlayColor(getContext()));

        DB.getInstance().hotkeyEntityDao().getAllObservable().observe(
                this, new Observer<List<HotkeyEntity>>() {
                    @Override
                    public void onChanged(List<HotkeyEntity> hotkeys) {
                        Log.v(TAG, "onChanged() for hotkeys");

                        if (hotkeys == null || hotkeys.isEmpty()) {
                            Log.d(TAG, "No hotkeys");
                            return;
                        }

                        uiHotkeysContainer.removeAllViews();

                        final int hotkeysContainerWidth = uiHotkeysContainer.getWidth();
                        for (HotkeyEntity hotkey : hotkeys) {
                            Log.i(TAG, "Hotkey " + hotkey.toString());
                            int xAbs = (int) (hotkeysContainerWidth * hotkey.xRel);

                            TextView h = newTextView(
                                    String.valueOf(hotkey.id),
                                    hotkey.name,
                                    hotkey.textSize,
//                                    hotkey.textColor,
//                                    hotkey.backgroundColor,
//                                    hotkey.borderSize,
//                                    hotkey.borderColor,
                                    hotkey.padding,
                                    xAbs,
                                    (int) hotkey.yAbs
                            );
                            uiHotkeysContainer.addView(h);
                        }

                        if (mSavedState != null) {
                            Log.d(TAG, "Restoring from a saved instance state, updating positions");

                            for (int i = 0; i < uiHotkeysContainer.getChildCount(); i++) {
                                View hotkeyView = uiHotkeysContainer.getChildAt(i);

                                if (!(hotkeyView instanceof TextView)) {
                                    Log.w(TAG, "No text view?");
                                    continue;
                                }

                                FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) hotkeyView.getLayoutParams();

                                final int id = Integer.valueOf((String) hotkeyView.getTag());

                                int[] viewPos = mSavedState.getIntArray(String.valueOf(id));
                                if (viewPos != null) {
                                    Log.d(TAG, "Restoring pos for [" + id + "]");
                                    lp.leftMargin = viewPos[0];
                                    lp.topMargin = viewPos[1];
                                    hotkeyView.setLayoutParams(lp);
                                }
                            }
                        }

//                        mSavedState = null;
                    }
                });
        setToolbarTitle("Hotkeys");
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mSavedState = saveState();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBundle(TAG, (mSavedState != null) ? mSavedState : saveState());
    }

    private Bundle saveState() {
        Bundle bundle = new Bundle();

        Log.d(TAG, "Fragment will go away, saving state");

        for (int i = 0; i < uiHotkeysContainer.getChildCount(); i++) {
            View hotkeyView = uiHotkeysContainer.getChildAt(i);

            if (!(hotkeyView instanceof TextView)) {
                Log.w(TAG, "No text view?");
                continue;
            }

            FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) hotkeyView.getLayoutParams();

            final int id = Integer.valueOf((String) hotkeyView.getTag());
            final int x = lp.leftMargin;
            final int y = lp.topMargin;

            Log.d(TAG, "Saving state of [" + id + "]");
            bundle.putIntArray(String.valueOf(id), new int[] {x, y});
        }

        return bundle;
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.hotkeys_menu, menu);
        mSaveButton = menu.findItem(R.id.uiSaveHotkeysMenuItem);
        mSaveButton.getIcon().mutate();
        mSaveButton.setEnabled(false);
        mSaveButton.getIcon().mutate().setAlpha(130);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.uiAddHotkeyMenuItem:
                handleAddButtonClick();
                return true;
            case R.id.uiSaveHotkeysMenuItem:
                handleSaveButtonClick();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private TextView newTextView(
            final String id,
            String name,
            int textSize,
//            int textColor,
//            int backgroundColor,
//            int borderSize,
//            int borderColor,
            int padding,
            int x, int y

    ) {

        final TextView textView = new TextView(getContext());
        textView.setText(name);
        textView.setTextSize(textSize);
//        textView.setTextColor(Color.RED);
//        textView.setBackgroundColor(backgroundColor);
        FrameLayout.LayoutParams lp = new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT
        );
        lp.leftMargin = x;
        lp.topMargin = y;
        textView.setLayoutParams(lp);
        textView.setPadding(padding, padding, padding, padding);

        textView.setBackground(ContextCompat.getDrawable(getContext(), R.drawable.touchpad_button_selector));

        GradientDrawable hotkeyBackground = new GradientDrawable();
        hotkeyBackground.setStroke(3, SettingsManager.hotkeyBorderColor(getContext()));
        hotkeyBackground.setCornerRadius(10);
        hotkeyBackground.setColor(SettingsManager.hotkeyBackgroundColor(getContext()));

        GradientDrawable hotkeyPressedBackground = new GradientDrawable();
        hotkeyPressedBackground.setStroke(3, SettingsManager.hotkeyBorderColor(getContext()));
        hotkeyPressedBackground.setCornerRadius(10);
        hotkeyPressedBackground.setColor(SettingsManager.hotkeyPressedBackgroundColor(getContext()));


        StateListDrawable hotkeySelector = new StateListDrawable();
        hotkeySelector.addState(new int[] {android.R.attr.state_pressed}, hotkeyPressedBackground);
        hotkeySelector.addState(new int[] {}, hotkeyBackground);

        textView.setBackground(hotkeySelector);
        textView.setTextColor(SettingsManager.hotkeyTextColor(getContext()));

        textView.setTag(id);

        textView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Log.v(TAG, "click");
                NavHostFragment.findNavController(HotkeysFragment.this)
                        .navigate(HotkeysFragmentDirections.actionAddEditHotkey(Integer.valueOf(id)));
            }
        });

        textView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                Log.v(TAG, "long click");

                // Create a new ClipData.
                // This is done in two steps to provide clarity. The convenience method
                // ClipData.newPlainText() can create a plain text ClipData in one step.

                // Create a new ClipData.Item from the ImageView object's tag
                ClipData.Item item = new ClipData.Item((v.getTag().toString()));

                // Create a new ClipData using the tag as a label, the plain text MIME type, and
                // the already-created item. This will create a new ClipDescription object within the
                // ClipData, and set its MIME type entry to "text/plain"
                ClipData dragData = new ClipData(
                        v.getTag().toString(),
                        new String[] { ClipDescription.MIMETYPE_TEXT_PLAIN },
                        item);

                // Instantiates the drag shadow builder.
//                View.DragShadowBuilder myShadow = new MyDragShadowBuilder(textView);
                View.DragShadowBuilder myShadow = new View.DragShadowBuilder(textView);

                v.startDragAndDrop(dragData, myShadow, null, 0);

                return true;
            }
        });

        uiHotkeysContainer.setOnDragListener(new View.OnDragListener() {
            @Override
            public boolean onDrag(View v, DragEvent event) {
                switch(event.getAction()) {
                    case DragEvent.ACTION_DRAG_STARTED:
                        Log.d(TAG, "Action is DragEvent.ACTION_DRAG_STARTED");

                        // Do nothing
                        break;

                    case DragEvent.ACTION_DRAG_ENTERED:
                        Log.d(TAG, "Action is DragEvent.ACTION_DRAG_ENTERED");
                        break;

                    case DragEvent.ACTION_DRAG_EXITED :
                        Log.d(TAG, "Action is DragEvent.ACTION_DRAG_EXITED");

                        break;

                    case DragEvent.ACTION_DRAG_LOCATION  :
                        Log.d(TAG, "Action is DragEvent.ACTION_DRAG_LOCATION");

                        break;

                    case DragEvent.ACTION_DRAG_ENDED   :
                        Log.d(TAG, "Action is DragEvent.ACTION_DRAG_ENDED");

                        // Do nothing
                        break;

                    case DragEvent.ACTION_DROP:
                        Log.d(TAG, "ACTION_DROP event for v: " + v);
                        ClipData.Item item = event.getClipData().getItemAt(0);
                        String dragData = item.getText().toString();
                        Log.d(TAG, "ACTION_DROP event for data: " + dragData);

                        TextView draggedTextView = uiHotkeysContainer.findViewWithTag(dragData);

                        int whalf = draggedTextView.getWidth() / 2;
                        int hhalf = draggedTextView.getHeight() / 2;
                        int x = (int) event.getX();
                        int y = (int) event.getY();
                        ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) draggedTextView.getLayoutParams();
                        lp.leftMargin = x - whalf;
                        lp.topMargin = y - hhalf;
                        draggedTextView.setLayoutParams(lp);
                        mSaveButton.setEnabled(true);
                        mSaveButton.getIcon().setAlpha(255);
                        break;
                    default:
                        break;
                }
                return true;
            }
        });

        return textView;
    }

    private void handleAddButtonClick() {
        Log.v(TAG, "Clicked on add button");
        NavHostFragment.findNavController(this).navigate(HotkeysFragmentDirections
                .actionAddEditHotkey(AddEditHotkeyFragment.ADD_HOTKEY_MAGIC_ACTIOPN_ID));
    }

    private void handleSaveButtonClick() {
        Log.v(TAG, "Clicked on save button, saving " + uiHotkeysContainer.getChildCount() + " views");
        for (int i = 0; i < uiHotkeysContainer.getChildCount(); i++) {
            View hotkeyView = uiHotkeysContainer.getChildAt(i);

            if (!(hotkeyView instanceof TextView)) {
                Log.w(TAG, "No text view?");
                continue;
            }

            FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) hotkeyView.getLayoutParams();

            final int id = Integer.valueOf((String) hotkeyView.getTag());
            // Save the X coordinate as relative accordingly to the width of th
            // e parent
            // In this way the disposition of the views is robust to the real container
            // width of the overlay on the controller
            final int xAbs = lp.leftMargin;
            final int yAbs = lp.topMargin;

            final double xRel = (double) xAbs / uiHotkeysContainer.getWidth();

            DB.getInstance().execute(new Runnable() {
                @Override
                public void run() {
                    Log.v(TAG, "Updating view [" + id + "] with pos X: " + xRel + ", Y: " + yAbs);
                    DB.getInstance().hotkeyEntityDao().updatePosition(id, xRel, yAbs);
                }
            });
        }

        goBack();
    }
}

File: app/src/main/java/org/docheinstein/minimotek/ui/hotkeys/HotkeysFragment.kt
package org.docheinstein.minimotek.ui.hotkeys

import android.content.ClipData
import android.os.Bundle
import android.view.*
import android.widget.FrameLayout
import android.widget.TextView
import androidx.core.view.children
import androidx.core.view.updateLayoutParams
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import dagger.hilt.android.AndroidEntryPoint
import org.docheinstein.minimotek.R
import org.docheinstein.minimotek.database.hotkey.Hotkey
import org.docheinstein.minimotek.databinding.HotkeysBinding
import org.docheinstein.minimotek.util.debug
import org.docheinstein.minimotek.util.warn
import kotlin.math.roundToInt

@AndroidEntryPoint
class HotkeysFragment : Fragment() {
    private val viewModel: HotkeysViewModel by viewModels()
    private lateinit var binding: HotkeysBinding

    private lateinit var saveButton: MenuItem

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        binding = HotkeysBinding.inflate(inflater, container, false)

        // Observe hw hotkeys list changes
        viewModel.hotkeys.observe(viewLifecycleOwner) { hotkeys ->
            if (hotkeys != null) {
                handleHotkeysUpdate(hotkeys)
            }
        }

        binding.hotkeys.setOnDragListener { v, e ->
            handleHotkeyDragged(v, e)
            true
        }

        return binding.root
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.add_save, menu)
        saveButton = menu.findItem(R.id.save_menu_item)
        saveButton.icon.mutate()
//        saveButton.isEnabled = false
//        saveButton.icon.mutate().alpha = 120
        setSaveButtonEnabled(false)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.add_menu_item -> {
                handleAddHotkeyButton()
                return true
            }
            R.id.save_menu_item -> {
                handleSaveHotkeysButton()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
//        when(item.itemId) {
//            R.id.edit_menu_item -> {
//                val hwHotkey = adapter.selected()
//                debug("Going to edit hwHotkey at position ${adapter.selection}: ${hwHotkey?.id}")
//                if (hwHotkey != null) {
//                    findNavController().navigate(
//                        HwHotkeysFragmentDirections.actionAddEditHwHotkey(
//                            hwHotkey.id,
//                            getString(R.string.toolbar_title_edit_hw_hotkey)
//                        )
//                    )
//                }
//            }
//            R.id.delete_menu_item -> {
//                val hwHotkey = adapter.selected()
//                debug("Going to delete hwHotkey at position ${adapter.selection}: ${hwHotkey?.id}")
//                if (hwHotkey != null) {
//                    AlertDialog.Builder(requireActivity())
//                        .setTitle(R.string.delete_hw_hotkey_confirmation_title)
//                        .setMessage(R.string.delete_hw_hotkey_confirmation_message)
//                        .setPositiveButton(R.string.ok) { _, _ ->
//                            // actually delete
//                            viewModel.delete(hwHotkey)
//                            Snackbar.make(
//                                requireParentFragment().requireView(),
//                                getString(R.string.hw_hotkey_removed, hwHotkey.button.name),
//                                Snackbar.LENGTH_LONG
//                            ).show()
//
//                            findNavController().navigateUp()
//                        }
//                        .setNegativeButton(R.string.cancel, null)
//                        .show()
//                }
//            }
//        }
        return super.onContextItemSelected(item)
    }

    private fun handleHotkeysUpdate(hotkeys: List<Hotkey>) {
        debug("Hotkey list has been updating, updating UI")

        binding.hotkeys.removeAllViews()

        for (h in hotkeys) {
            val hotkeyView = makeHotkeyView(h)
            binding.hotkeys.addView(hotkeyView)
        }
    }


    private fun handleHotkeyDragged(v: View, ev: DragEvent) {
        when (ev.action) {
            DragEvent.ACTION_DRAG_STARTED -> {
                debug("Drag event detected: ACTION_DRAG_STARTED")
            }
            DragEvent.ACTION_DRAG_ENTERED -> {
                debug("Drag event detected: ACTION_DRAG_ENTERED")
            }
            DragEvent.ACTION_DRAG_EXITED -> {
                debug("Drag event detected: ACTION_DRAG_EXITED")
            }
            DragEvent.ACTION_DRAG_LOCATION -> {
                debug("Drag event detected: ACTION_DRAG_LOCATION")
            }
            DragEvent.ACTION_DRAG_ENDED -> {
                debug("Drag event detected: ACTION_DRAG_ENDED")
            }
            DragEvent.ACTION_DROP -> {
                debug("Drag event detected: ACTION_DROP")
                if (ev.clipData.itemCount < 1) {
                    warn("Invalid event data")
                    return
                }

                val hotkeyIdStr: String = ev.clipData.getItemAt(0).text.toString()

                debug("Dropped hotkey: $hotkeyIdStr")

                val hotkeyView: HotkeyView? = binding.hotkeys.findViewWithTag(hotkeyIdStr)

                if (hotkeyView == null) {
                    warn("Failed to find hotkey for tag $hotkeyIdStr")
                    return
                }

                val halfWidth = hotkeyView.width / 2
                val halfHeight = hotkeyView.height / 2

                hotkeyView.updateLayoutParams<ViewGroup.MarginLayoutParams> {
                    val x = (ev.x - halfWidth).toInt()
                    val y = (ev.y - halfHeight).toInt()
                    debug("Updating X = $x, Y = $y")
                    this.leftMargin = x
                    this.topMargin = y
                }

                setSaveButtonEnabled(true)
            }
            else -> {
                warn("Unknown rag event detected")
            }
        }
    }

    private fun makeHotkeyView(hotkey: Hotkey): View {
        val hotkeyIdStr = hotkey.id.toString()

        val hotkeyView = HotkeyView(requireContext(), hotkey = hotkey)
        hotkeyView.tag = hotkeyIdStr

        val lp = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        )
        lp.leftMargin = hotkey.x
        lp.topMargin = hotkey.y
        hotkeyView.layoutParams = lp

        hotkeyView.setOnClickListener {
            findNavController().navigate(
                HotkeysFragmentDirections.actionAddEditHotkey(
                    hotkey.id,
                    getString(R.string.toolbar_title_edit_hotkey)
                )
            )
        }

        hotkeyView.setOnLongClickListener {
            debug("Long click on hotkey $hotkey")

            it.startDragAndDrop(
                ClipData.newPlainText(hotkeyIdStr, hotkeyIdStr),
                View.DragShadowBuilder(hotkeyView),
                hotkey, 0
            )

            true
        }

        return hotkeyView
    }

    private fun handleSaveHotkeysButton() {
        debug("Clicked on save, saving ${binding.hotkeys.childCount} hotkeys")

        for (hotkeyView in binding.hotkeys.children) {
            if (hotkeyView !is HotkeyView) {
                warn("Child view is not an hotkey view!?")
                continue
            }

            val lp = hotkeyView.layoutParams as FrameLayout.LayoutParams
            val id = (hotkeyView.tag as String).toLong()
            val x = lp.leftMargin
            val y = lp.topMargin

            debug("Updating hotkey $id to position ($x,$y)")
            viewModel.updatePosition(id, x, y)
        }

        findNavController().navigateUp()
    }

    private fun handleAddHotkeyButton() {
        findNavController().navigate(
            HotkeysFragmentDirections.actionAddEditHotkey(
                AddEditHotkeyViewModel.HOTKEY_ID_NONE,
                getString(R.string.toolbar_title_add_hotkey)
            )
        )
    }

    private fun setSaveButtonEnabled(enabled: Boolean) {
        if (enabled) {
            saveButton.isEnabled = true
            saveButton.icon.alpha = 255
        } else {
            saveButton.isEnabled = false
            saveButton.icon.alpha = 127
        }
    }
}